src/test/java/com/shapesecurity/shift/parser/RangeCheckerReducer.java

```diff
@@ -0,0 +1,711 @@
+/*
+ * Copyright 2014 Shape Security, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"));
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.shapesecurity.shift.parser;
+
+import com.shapesecurity.functional.data.ImmutableList;
+import com.shapesecurity.functional.data.Maybe;
+import com.shapesecurity.functional.data.Monoid;
+import com.shapesecurity.shift.ast.*;
+import com.shapesecurity.shift.visitor.MonoidalReducer;
+import org.jetbrains.annotations.NotNull;
+
+import static org.junit.Assert.assertTrue;
+
+public class RangeCheckerReducer extends MonoidalReducer<RangeCheckerReducer.RangeChecker> {
+	private final ParserWithLocation parserWithLocation;
+
+	protected RangeCheckerReducer(ParserWithLocation parserWithLocation) {
+		super(RangeChecker.MONOID);
+		this.parserWithLocation = parserWithLocation;
+	}
+
+	private RangeChecker accept(Node node, RangeChecker innerBounds) {
+		Maybe<SourceSpan> span = this.parserWithLocation.getLocation(node);
+		assertTrue(span.isJust());
+		RangeChecker outerBounds = new RangeChecker(span.just());
+		assertTrue(outerBounds.start <= outerBounds.end);
+
+		assertTrue(outerBounds.start <= innerBounds.start);
+		assertTrue(innerBounds.end <= outerBounds.end);
+
+		return outerBounds;
+	}
+
+	@NotNull
+	@Override
+	public RangeChecker reduceArrayBinding(
+		@NotNull ArrayBinding node, @NotNull ImmutableList<Maybe<RangeChecker>> elements,
+		@NotNull Maybe<RangeChecker> restElement
+	) {
```

Comment:
We don't really have a consistent style in this repo, and it was way better than what IntelliJ was giving me by default.